src/wit_gen.rs

```diff
@@ -228,4 +307,121 @@ mod tests {
+        });
+
+        let generator = AutoTraceGenerator {
+            auto_signals: HashMap::from([(step_uuid, HashMap::from([(c, a + b)]))]),
+        };
+
+        let witness = generator.generate(witness);
+        assert_eq!(
+            witness.step_instances[0].assignments.get(&c),
+            Some(&Fr::from(2))
+        );
+        assert_eq!(
+            witness.step_instances[1].assignments.get(&c),
+            Some(&Fr::ONE)
+        );
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_auto_trace_gen_panic() {
+        let a = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "a".to_string()),
+            false,
+        );
+        let b = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "b".to_string()),
+            false,
+        );
+        let c = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "c".to_string()),
+            false,
+        );
+        let step_uuid = uuid();
+        let mut witness = TraceWitness::default();
+        witness.step_instances.push(StepInstance {
+            step_type_uuid: step_uuid,
+            assignments: HashMap::from([(a, Fr::ONE)]),
+        });
+
+        let generator = AutoTraceGenerator {
+            auto_signals: HashMap::from([(step_uuid, HashMap::from([(c, a + b)]))]),
+        };
+
+        generator.generate(witness);
+    }
+
+    #[test]
+    fn test_auto_trace_gen_dep() {
+        let a = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "a".to_string()),
+            false,
+        );
+        let b = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "b".to_string()),
+            false,
+        );
+        let c = Queriable::Forward(
+            ForwardSignal::new_with_id(uuid(), 0, "c".to_string()),
+            false,
+        );
+        let step_uuid = uuid();
+        let mut witness = TraceWitness::default();
+        witness.step_instances.push(StepInstance {
+            step_type_uuid: step_uuid,
+            assignments: HashMap::from([(a, Fr::ONE)]),
+        });
+        witness.step_instances.push(StepInstance {
+            step_type_uuid: step_uuid,
+            assignments: HashMap::from([(a, Fr::ONE), (c, Fr::ONE)]),
+        });
+
+        let generator = AutoTraceGenerator {
+            auto_signals: HashMap::from([(step_uuid, HashMap::from([(c, a + b), (b, a + 1)]))]),
+        };
+
+        let witness = generator.generate(witness);
+        assert_eq!(
+            witness.step_instances[1].assignments.get(&b),
+            Some(&Fr::from(2))
+        );
+        assert_eq!(
+            witness.step_instances[0].assignments.get(&c),
+            Some(&Fr::from(3))
+        );
+        assert_eq!(
+            witness.step_instances[1].assignments.get(&b),
+            Some(&Fr::from(2))
+        );
+        assert_eq!(
+            witness.step_instances[1].assignments.get(&c),
+            Some(&Fr::ONE)
+        );
```

Comment:
Looks like this example is trying to test that if there's an existing assignment for `c = 1`, then the auto-generator won't mask it with inferred value `c = 3`.

I think you called `assert_eq!` twice on the same assignment `witness.step_instances[1].assignments.get(&b)`, so there's probably something else you are trying to assert.

Reply:
This was an error in the tests, the duplicated assert_eq!, it is fixed now

Reply:
LGTM.